netket/logging/tar_log.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import dataclasses
+import tarfile
+import time
+from io import BytesIO
+
+import os
+from os import path as _path
+import numpy as np
+import jax
+
+from flax import serialization
+
+
+def save_binary_to_tar(tar_file, byte_data, name):
+    abuf = BytesIO(byte_data)
+
+    # Contruct the info object with the correct length
+    info = tarfile.TarInfo(name=name)
+    info.size = len(abuf.getbuffer())
+
+    # actually save the data to the tar file
+    tar_file.addfile(tarinfo=info, fileobj=abuf)
+
+
+class TarLog:
+    """
+        Tar Logger, serializing the variables of the variational state during a run.
+
+    : bool flag indicating whever to store variables in a tar file. The tar archive will
+                    contain a file with numbers going from 0 to N, and every file corresponds to the variables of
+                    the variational state at that step.
+
+
+        Data is serialized to a tar archive as many files named `[0.mpack, 1.mpack, ...]`, where
+        the number increases by 1 every time it's called.
+
+        The tar file inside is not flushed to disk (closed) until this object is deleted or python
+        is shut down.
```

Comment:
Is this save in case the simulation run crashes? Being able to recover from interruptions seems like an important use case for storing intermediate states on disk. If an improper shutdown of the process can leave the tar file in a corrupted state this would not work.

(Also, being able to already access the intermediate states in a long running simulation can be pretty useful by itself.)

Reply:
This does close the file handle if an exception is thrown (it doesn't if there is a segmentation fault, but in that case it's impossible to recover).

However, to make him 'flush' the file I would have to open and close the file handle after every write. Closing is cheap (he just writes a small ending) but reopening is more expensive as he repress the index of thins inside the tar file.

I think it's not worth doing it.

Reply:
Hm, in that case I'd prefer an option of just storing the files one by one in a directory (where this is not an issue as a crash would only affect the last file if one is unlucky).

But that's maybe separate from this PR.

Reply:
maybe we can rename to StateLog and have an option `tar=True/False`?

Reply:
I think that would be pretty useful!